wiki_articles/std_array.md

```diff
@@ -0,0 +1,33 @@
+# What Is `std::array` And Why Should I Use It?
+
+C-style arrays have a few properties that make them annoying to use. They:
+
+:no_entry: cannot be returned from functions<br>
+:no_entry: are non-assignable<br>
+:no_entry: can only be initialized via `""`, `{}`, or *[default initialization](https://en.cppreference.com/w/cpp/language/default_initialization)*<br>
+:no_entry: cannot be passed to functions by value, but *[decay](https://64.github.io/cpp-faq/decay/)* to pointers<br>
+:warning: might implicitly turn into pointers, which is called *[array-to-pointer conversion](https://en.cppreference.com/w/cpp/language/implicit_conversion#Array-to-pointer_conversion)*<br>
+:warning: might be [variable-length arrays](https://en.wikipedia.org/wiki/Variable-length_array) (VLAs), if the developer makes a mistake
+
+**[std::array](https://en.cppreference.com/w/cpp/container/array)** is a standard library
+*[SequenceContainer](https://en.cppreference.com/w/cpp/named_req/SequenceContainer)* and
```

Comment:
It'd probably be better to leave off SequenceContainer here, since this article is geared towards beginners. Maybe not aggregate type either. Similarly I don't think it's worth including implementation details, the example usage is more helpful.

Reply:
I agree that maybe including SequenceContainer is a bit too much, but the rest should stay. The easiest way to explain `std::array` is by showing people that it's basically just a C-style array wrapped in a struct, not some complicated construct.

I see that explanation being thrown around a lot, and also there is no way around it when running into brace dilemmas like:
```cpp
std::array<std::array<int, 2> 2> arr{{1, 2}, {3, 4}}; // ill-formed
```
You cannot do that because `{1, 2}` initializes the C-style array inside of the outer `std::array` and `{3, 4}` is stray. There's basically no way to make sense of this unless you know about the implementation, and this also makes it important to know that `std::array` is an aggregate type.

We could handwave this issue and tell beginners to use brace elision like `{1, 2, 3, 4}`, but explaining brace elision is a bigger can of worms, and this advice would be very exotic.

Reply:
I'm good with keeping the comment about it being an aggregate type, that's a helpful term for beginners to get familiar with